src/value/borrowed.rs

```diff
@@ -0,0 +1,229 @@
+}
+
+impl<'val> BorrowedSequence<'val> {
+    pub fn new(children: Vec<BorrowedElement<'val>>) -> Self {
+        Self { children }
+    }
+}
+
+impl<'val> Sequence for BorrowedSequence<'val> {
+    type Element = BorrowedElement<'val>;
+
+    fn iter<'a>(&'a self) -> Box<dyn Iterator<Item = &'a Self::Element> + 'a> {
+        Box::new(self.children.iter())
+    }
+}
+
+/// A borrowed implementation of [`Struct`]
+#[derive(Debug, Clone)]
+pub struct BorrowedStruct<'val> {
+    // TODO model actual map indexing for the struct (maybe as a variant type)
+    //      otherwise struct field lookup will be O(N)
+    fields: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>,
+}
+
+impl<'val> BorrowedStruct<'val> {
+    pub fn new(fields: Vec<(BorrowedSymbolToken<'val>, BorrowedElement<'val>)>) -> Self {
+        Self { fields }
+    }
+}
+
+impl<'val> Struct for BorrowedStruct<'val> {
+    type FieldName = BorrowedSymbolToken<'val>;
+    type Element = BorrowedElement<'val>;
+
+    fn iter<'a>(
+        &'a self,
+    ) -> Box<dyn Iterator<Item = (&'a Self::FieldName, &'a Self::Element)> + 'a> {
+        // convert &(k, v) -> (&k, &v)
+        Box::new(self.fields.iter().map(|kv| match &kv {
+            (k, v) => (k, v),
+        }))
+    }
+}
+
+// TODO replace the references with `Cow` and bridge to the owned APIs for mutability
+
+/// Variants for all borrowed version _values_ within an [`Element`].
+#[derive(Debug, Clone)]
+pub enum BorrowedValue<'val> {
+    Null(IonType),
+    Integer(AnyInt),
+    String(&'val str),
+    Symbol(BorrowedSymbolToken<'val>),
+    SExpression(BorrowedSequence<'val>),
+    List(BorrowedSequence<'val>),
+    Struct(BorrowedStruct<'val>),
+    // TODO fill this in with the rest...
+}
+
+/// A borrowed implementation of [`Element`]
+#[derive(Debug, Clone)]
+pub struct BorrowedElement<'val> {
+    annotations: Vec<BorrowedSymbolToken<'val>>,
+    value: BorrowedValue<'val>,
+}
+
+impl<'val> BorrowedElement<'val> {
+    pub fn new(annotations: Vec<BorrowedSymbolToken<'val>>, value: BorrowedValue<'val>) -> Self {
+        Self { annotations, value }
+    }
+}
+
+impl<'val> From<BorrowedValue<'val>> for BorrowedElement<'val> {
+    /// Constructs a [`BorrowedElement`] without annotations from this value.
+    fn from(val: BorrowedValue<'val>) -> Self {
+        Self::new(vec![], val)
+    }
+}
+
+impl<'val> Element for BorrowedElement<'val> {
+    type SymbolToken = BorrowedSymbolToken<'val>;
+    type Sequence = BorrowedSequence<'val>;
+    type Struct = BorrowedStruct<'val>;
+
+    fn ion_type(&self) -> IonType {
+        use BorrowedValue::*;
+        match &self.value {
+            Null(t) => *t,
+            Integer(_) => IonType::Integer,
+            String(_) => IonType::String,
+            Symbol(_) => IonType::Symbol,
+            SExpression(_) => IonType::SExpression,
+            List(_) => IonType::List,
+            Struct(_) => IonType::Struct,
+        }
+    }
```

Comment:
Does this effectively strip out type information from null values?

(I chose to encode each of the underlying values as an `Option<T>` which worked pretty well, didn't lose the type info, and made some later matching really easy)

Reply:
No, this explicitly models it as the `Null(IonType)` variant so one does not need to have `Option<V>` for all the other variants.

Reply:
Ah, I see it now. I just read this wrong and didn't pay close enough attention to the variant definition. Nice.